### PR TITLE
feat: Add hideSchemas swagger option (#686)

### PR DESCRIPTION
**Pull request by imjuni**
* feat: add hideSchemas swagger option
* test: add tests for hideSchemas swagger option
* docs: add hideSchemas swagger option

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/spec/swagger/schema.js [hunk 1]
```diff
@@ -481,3 +481,74 @@ test('support "const" keyword', async t => {
...
+    schema: {
+      response: {
+        200: {
+          type: 'object'
+        }
+      }
+    }
+  }
+
+  fastify.get('/', opts9, () => {})
```
Comment (mcollina) [inline test/spec/swagger/schema.js:527]: Can you please add a test that uses `schema-02` as a ref?

File: README.md [hunk 2]
```diff
@@ -219,6 +219,7 @@ An example of using `@fastify/swagger` with `static` mode enabled can be found [
  | swagger            | {}               | [Swagger configuration](https://swagger.io/specification/v2/#swaggerObject).                                              |
  | transform          | null             | Transform method for the route's schema and url. [documentation](#register.options.transform).                                                                                              |
  | refResolver        | {}               | Option to manage the `$ref`s of your application's schemas. Read the [`$ref` documentation](#register.options.refResolver) |
+ | hideSchemas        | []               | Option to control hiding of schemas.                                                                                       |
```
Comment (Uzlopak) [inline README.md:222]: I dont get it. Hiding of Schemas from who or where?

File: test/spec/swagger/schema.js (hunk 1 above)
Comment (imjuni) [reply test/spec/swagger/schema.js]: Thank you for your review.

I change code like that,

```ts
fastify.get('/', {
    schema: {
      querystring: { $ref: 'schema-02' },
      response: {
        200: {
          type: 'object'
        }
      }
    }
  }, () => {})
```

File: README.md (hunk 2 above)
Comment (imjuni) [reply README.md]: Thank you for your review,

Change description like that,

```
Option to control hiding of schemas in swagger document definitions
```

How about this? I mean only hide that from definition sections in swagger documents.

File: test/spec/openapi/schema.js [hunk 3]
```diff
@@ -849,3 +849,73 @@ test('support query serialization params', async t => {
...
+          type: 'object'
+        }
+      }
+    }
+  }, () => {})
+
+  await fastify.ready()
+
+  const openapiObject = fastify.swagger()
+  await Swagger.validate(openapiObject)
```
Comment (mcollina) [inline test/spec/openapi/schema.js:899]: I don't understand how this would be valid.

Comment (climba03003) [reply test/spec/openapi/schema.js:899]: `Swagger.validate` only validate `structure`. It does not validate the linkage.
Yes, if `schema-02` is hidden, it should not be a valid one.

**Review by mcollina — APPROVED** [review]
lgtm

File: test/spec/openapi/schema.js (hunk 3 above)
Comment (imjuni) [reply test/spec/openapi/schema.js:899]: I think that difference between json schema store and swagger document definition. Validation need every schema but swagger document don't have to display every definitions.

Also, I can move the logic of hiding the schema to the last time of the document building like OpenAPI v2.

**Comment by Uzlopak** [comment]
I am now too lazy to read the code. But the only way to avoid those schemas to be rendered in the final document should be by resolving the references with the schemas.

File: lib/spec/openapi/index.js [hunk 4]
```diff
@@ -20,9 +20,12 @@ module.exports = function (opts, cache, routes, Ref, done) {
     const openapiObject = prepareOpenapiObject(defOpts, done)
 
     ref = Ref()
+    const definitions = Object.entries(ref.definitions().definitions)
+      .filter(([, schema]) => defOpts.hideSchemas.findIndex(hideSchema => hideSchema === schema.$id) < 0)
+      .reduce((defs, [$id, schema]) => ({ ...defs, [$id]: schema }), {})
```
Comment (climba03003) [inline lib/spec/openapi/index.js:25]: Can you change this code into a single `for loop`?
It repeatedly cloning the `defs`.

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: lib/spec/openapi/index.js (hunk 4 above)
Comment (imjuni) [reply lib/spec/openapi/index.js]: Sure, I chaged it.

**Review by climba03003 — APPROVED** [review]

**Comment by imjuni** [comment]
@Uzlopak , @mcollina Here is example, 

```js
const fastify = require('fastify');
const server = fastify();

// case 01. error raise when server start
server.addSchema({
  $id: 'IReqPokeDetailQuerystring',
  type: 'object',
  properties: {
    tid: {
      $ref: 'missing reference!',
    },
  },
  required: ['tid'],
});

// case 02. error raise when request call
server.get(
  '/pokemon/:name',
  {
    schema: {
      querystring: {
        type: 'object',
        properties: {
          tid: {
            $ref: 'missing reference!',
          },
        },
        required: ['tid'],
      },
    },
  },
  () => {
    /* ... */
  },
);
```

I prefer what case01 then case02. Because more than quickly aware error from validation schema. But case01 generate so many DTO in swagger document. My company project already over 100 DTOs. Since Querystirng, Headers, Params, and Body each generate a DTO, the DTO can be increased very much. So I hope hide Querystring, Headers, Params and Body DTO. Because Querystring, Headers, Params and Body already describe in API spec.

**Comment by Uzlopak** [comment]
I understand that, but you need to resolve the references. If the DTO is not used anyway, it should not be added to the final swagger document.

But you are just manually suppress the propagation to the final document.

So e.g. i would actually expect that if you specifiy IReqPokeDetailQuerystring as a reference that if it is used somewhere to be resolved so that the specification is still valid and remove it from the final swagger documentation.  If it is not referenced anywhere, it would be anyway removed. 

So instead of hideSchemas i would expect a resolveReferences option, where you pass an Array or maybe  a function where the name of a reference is passed and then decided if it should be resolved or not

```typescript
resolveReference: (name: string) => boolean
```

Then you could do something like

```js
function resolveReference(name) {
   return name.endsWith("Querystring") === false
}
```

**Comment by imjuni** [comment]
I understand resolve reference. But some difference of this case. First I try to remove reference like that,

```js
ref = Ref()
const prevDefinitions = prepareSwaggerDefinitions({
  ...swaggerObject.definitions,
  ...(ref.definitions().definitions)
}, ref)

// remove reference
const definitions = {}
for (const [$id, schema] of Object.entries(prevDefinitions)) {
  if (defOpts.resolveReference(schema.$id)) {
    definitions[$id] = schema
  }
}
```

But this approach fail to build document in case of OpenAPI v2. Because OpenAPI v2 use reference for document building. This work raise error from [prepareSwaggerMethod](https://github.com/fastify/fastify-swagger/blob/c8d675d819f7fd274f00a20144a4efdca0c55778/lib/spec/swagger/index.js#L46). But OpenAPI v3 fine work this approach. I thought it would be better for v2 and v3 to behave the same. Also I can chage option type `string[]` to `Function`. Like that, 

```js
const definitions = {}
for (const [$id, schema] of Object.entries(swaggerObject.definitions)) {
  if (defOpts.isHideReference(schema.title)) {
    definitions[$id] = schema
  }
}
```

**Comment by Uzlopak** [comment]
Imho you are just removing the DTOs from the final document.  I mean your unit tests actually show it beatifully. 
So despite you added schema-0, schema-1 and schema-2 you manually remove schema-1, and because you dont use it in your schema building the schema succeeds . If we would remove schema-2, we should get a FST_ERR_SCH_SERIALIZATION_BUILD-Error.

We should actually fist add a flag which removes unused definitions from the final swagger-documentation.

Example unit test:

```js
test('remove unused schemas', async t => {
  t.plan(2)
  const fastify = Fastify()
  await fastify.register(fastifySwagger, { ...openapiOption, removeUnusedSchemas: true })

  fastify.addSchema({
    $id: 'schema-01',
    type: 'object',
    properties: {
      id: { type: 'number', description: 'data id' }
    },
    required: ['id']
  })

  fastify.addSchema({
    $id: 'schema-02',
    type: 'object',
    properties: {
      id: { type: 'number', description: 'data id' }
    },
    required: ['id']
  })

  fastify.addSchema({
    $id: 'schema-03',
    type: 'object',
    properties: {
      id: { type: 'number', description: 'data id' }
    },
    required: ['id']
  })

  fastify.get('/', {
    schema: {
      querystring: { $ref: 'schema-02' },
      response: {
        200: {
          type: 'object'
        }
      }
    }
  }, () => {})

  await fastify.ready()

  const openapiObject = fastify.swagger()
  await Swagger.validate(openapiObject)

  const definitions = openapiObject.components.schemas
  t.ok(definitions)
  t.same(definitions, {
    'def-2': {
      type: 'object',
      properties: {
        id: { type: 'number', description: 'data id' }
      },
      required: ['id'],
      title: 'schema-03'
    }
  })
})
```

In a second step, we should implement a flag, which forces to resolve specified DTOs. And then remove them from being exposed in the final swagger document.

**Pull request closed without merging**